### ISSUE TITLE
EventHandler Leak: System.AssemblyLoadEventHandler . += Uno.Xaml.XamlSchemaContext

### ISSUE THREAD

**Issue by calvinhsia** [unoplatform/uno#7421]
### Current behavior

https://github.com/unoplatform/uno/issues/5284
Hi Jerome Laban

This is still happening in VS 2022 17.0.31717.71

I don't know how to tell Uno version:
![t](https://user-images.githubusercontent.com/15060143/139793926-f5bc9f81-edea-43a2-a014-bc0c2b045bab.png)

FileName                                                    Location                                                                                                                                                        IsManaged ImageBase          Version
Uno.Xaml.dll                                                C:\Users\JSinsel\AppData\Local\Temp\VS\AnalyzerAssemblyLoader\0a786622a2214cf3891a83eb834d520a\15\Uno.Xaml.dll                                                  True      0000021d`2a5d0000  255.255.255.255
Uno.Core.dll                                                C:\Users\JSinsel\AppData\Local\Temp\VS\AnalyzerAssemblyLoader\0a786622a2214cf3891a83eb834d520a\8\Uno.Core.dll                                                   True      0000021d`2a450000  1.0.0.0
Uno.Analyzers.dll                                           C:\Users\JSinsel\AppData\Local\Temp\VS\AnalyzerAssemblyLoader\0a786622a2214cf3891a83eb834d520a\3\Uno.Analyzers.dll                                              True      0000021d`3b010000  1.0.0.0
Uno.UI.SourceGenerators.dll                                 C:\Users\JSinsel\AppData\Local\Temp\VS\AnalyzerAssemblyLoader\0a786622a2214cf3891a83eb834d520a\29\Uno.UI.SourceGenerators.dll                                   True      0000021d`3f900000  255.255.255.255

**Comment by jeromelaban** [comment]
Thanks for the report, we'll look into it.

**Comment by jeromelaban** [comment]
@calvinhsia This looks like there's a memory leak somewhere else, as the this particular piece is managed in such a way that it won't stay in memory if the owner class is collected properly. 

I know there were in VS2022 where many source generators were executed in loops and kept alive for a long time, and I wonder if that may be the cause. VS2022 Preview 6/7 seem to have improved in this particular.

Can you tell me how many instances of the `Uno.UI.SourceGenerators.XamlGenerator.XamlFileGenerator` type are present in the process in this snapshot?

**Comment by jeromelaban** [comment]
Thanks for the details @calvinhsia, this is very useful.

For what I can see, this would look like an excessive registration to the the `AssemblyLoaded` event, not necessarily a leak per-say.  We'll make changes to reduce the number of registrations to the event, likely to be just one.

I assume that what you're looking at is a snapshot of the process, but it does not tell you if this is a leak over time, is this correct ?

**Comment by calvinhsia** [comment]
this is from a single dump, but it is evidence of a leak over time

**Comment by jeromelaban** [comment]
Understood. The reason I'm saying this is that the number of registrations to the even can be really high in normal conditions, even if there's no leak.

Either way, it's not correct, so we'll change it, thanks for the help!

**Comment by davkean** [comment]
I see this in https://developercommunity.visualstudio.com/t/Visual-Studio-2022-Freezing-when-canceli/1593670, however, I see another issue.

We have over 700 threads all stuck on this:

```
 	ntdll.dll!ZwWaitForMultipleObjects() Line 947	Unknown	Symbols loaded.
 	KERNELBASE.dll!WaitForMultipleObjectsEx()	Unknown	Symbols loaded.
 	[Managed to Native Transition]		Annotated Frame
 	mscorlib.dll!System.Threading.WaitHandle.InternalWaitOne(System.Runtime.InteropServices.SafeHandle waitableSafeHandle, long millisecondsTimeout, bool hasThreadAffinity, bool exitContext) Line 243	C#	Symbols loaded.
 	mscorlib.dll!System.Threading.WaitHandle.WaitOne(System.TimeSpan timeout, bool exitContext) Line 205	C#	Symbols loaded.
 	mscorlib.dll!System.Threading.WaitHandle.WaitOne(System.TimeSpan timeout) Line 225	C#	Symbols loaded.
>	Microsoft.ApplicationInsights.dll!Microsoft.ApplicationInsights.Channel.InMemoryTransmitter.Runner() Line 113	C#	Symbols loaded.
 	mscorlib.dll!System.Threading.Tasks.Task.Execute() Line 2498	C#	Symbols loaded.
 	mscorlib.dll!System.Threading.ExecutionContext.RunInternal(System.Threading.ExecutionContext executionContext, System.Threading.ContextCallback callback, object state, bool preserveSyncCtx) Line 980	C#	Symbols loaded.
 	mscorlib.dll!System.Threading.ExecutionContext.Run(System.Threading.ExecutionContext executionContext, System.Threading.ContextCallback callback, object state, bool preserveSyncCtx) Line 928	C#	Symbols loaded.
 	mscorlib.dll!System.Threading.Tasks.Task.ExecuteWithThreadLocal(ref System.Threading.Tasks.Task currentTaskSlot) Line 2827	C#	Symbols loaded.
 	mscorlib.dll!System.Threading.Tasks.Task.ExecuteEntry(bool bPreventDoubleExecution) Line 2767	C#	Symbols loaded.
 	mscorlib.dll!System.Threading.ExecutionContext.RunInternal(System.Threading.ExecutionContext executionContext, System.Threading.ContextCallback callback, object state, bool preserveSyncCtx) Line 980	C#	Symbols loaded.
 	mscorlib.dll!System.Threading.ExecutionContext.Run(System.Threading.ExecutionContext executionContext, System.Threading.ContextCallback callback, object state, bool preserveSyncCtx) Line 928	C#	Symbols loaded.
 	mscorlib.dll!System.Threading.ExecutionContext.Run(System.Threading.ExecutionContext executionContext, System.Threading.ContextCallback callback, object state) Line 917	C#	Symbols loaded.
 	mscorlib.dll!System.Threading.ThreadHelper.ThreadStart(object obj) Line 93	C#	Symbols loaded.
 	[Native to Managed Transition]		Annotated Frame
 	kernel32.dll!BaseThreadInitThunk(unsigned long RunProcessInit, long(*)(void *) StartAddress, void * Argument) Line 75	C	Symbols loaded.
 	ntdll.dll!RtlUserThreadStart(long(*)(void *) StartAddress, void * Argument) Line 1152	C	Symbols loaded.

```

That AppInsights dll has been loaded next to where Uno generators have been shadow copied. Which implies to me that this also leaking, but I don't know how this telemetry works.  Looks like something is not calling InMemoryTransmitter.Dispose to tear this thread down? Is this the same underlying issue or should we open a new one?

**Comment by jeromelaban** [comment]
We've made changes to this somewhat recently, so it may depend on which version developers use, we'll get this fixed as well. Apologies for the trouble it's causing you for searching for these.

**Comment by davkean** [comment]
Is the thread leak the same issue?

**Comment by jeromelaban** [comment]
This is a different issue, though we'll fix it at the same time by reducing the work during design-time builds.

**Comment by jeromelaban** [comment]
Both the new thread and even leak are fixed in Uno 4.0.7. Thanks both for your help on these issues!
